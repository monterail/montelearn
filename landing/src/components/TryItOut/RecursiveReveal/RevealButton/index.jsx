import { useCallback } from "react";

export default function RevealButton({ css, onClick, onClickData, ...props }) {
  const handleClick = useCallback(() => {
    onClick(onClickData);
  }, [onClick, onClickData]);

  return (
    <button
      data-testid="RevealButton"
      css={{
        MozAppearance: "none",
        WebkitAppearance: "none",
        appearance: "none",
        backgroundColor: "transparent",
        border: "none",
        cursor: "pointer",
        fontFamily: "inherit",
        margin: 0,
        padding: 0,

        ":hover": {
          textDecoration: "underline",
        },

        ...css,
      }}
      onClick={handleClick}
      type="button"
      {...props}
    >
      show/hide
    </button>
  );
}
