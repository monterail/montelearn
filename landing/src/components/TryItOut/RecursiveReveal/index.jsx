import { createContext, useCallback, useContext, useState } from "react";

import { COLOR_DON_JUAN, COLOR_GREEN } from "@/theming/const";
import { rem } from "@/theming/utils";
import { isUrlLike } from "@/utils/paths";
import { pluralize } from "@/utils/wording";

import RevealButton from "./RevealButton";

const Context = createContext();

function ValueReveal({ id, value }) {
  const { isActive, toggle } = useContext(Context);
  const expanded = isActive(id);

  if (value === null) {
    return <span data-testid="RecursiveReveal_Null">NULL</span>;
  }

  if (typeof value === "undefined") {
    return <span data-testid="RecursiveReveal_Undefined">undefined</span>;
  }

  if (Array.isArray(value)) {
    return (
      <>
        <span data-testid="RecursiveReveal_Array">
          <span css={{ color: COLOR_DON_JUAN }}>Array</span>{" "}
          <i css={{ fontSize: rem(10) }}>
            ({value.length > 0 ? pluralize(value.length, "element", "elements") : "empty"})
          </i>{" "}
          {value.length > 0 && (
            <RevealButton
              data-testid="RecursiveReveal_ArrayReveal"
              onClick={toggle}
              onClickData={id}
            />
          )}
        </span>
        {expanded && (
          <div data-testid="RecursiveReveal_ArrayProperties" css={{ marginLeft: "1em" }}>
            {value.map((el, idx) => (
              <div data-testid="RecursiveReveal_ArrayProperty" key={idx}>
                <span css={{ color: COLOR_GREEN }}>{idx}</span>:{" "}
                <ValueReveal id={`${id}.${idx}`} value={el} />
              </div>
            ))}
          </div>
        )}
      </>
    );
  }

  if (value instanceof Date) {
    return <span data-testid="RecursiveReveal_Date">Date ({value.toISOString()})</span>;
  }

  if (typeof value === "object") {
    const keys = Object.keys(value);
    const keysCount = keys.length;

    return (
      <>
        <span data-testid="RecursiveReveal_Object">
          <span css={{ color: COLOR_DON_JUAN }}>Object</span>{" "}
          <i css={{ fontSize: rem(10) }}>
            ({keysCount > 0 ? pluralize(keysCount, "key", "keys") : "empty"})
          </i>{" "}
          {keysCount > 0 && (
            <RevealButton
              data-testid="RecursiveReveal_ObjectReveal"
              onClick={toggle}
              onClickData={id}
            />
          )}
        </span>
        {expanded && (
          <div data-testid="RecursiveReveal_ObjectProperties" css={{ marginLeft: "1em" }}>
            {keys.map((key) => (
              <div data-testid="RecursiveReveal_ObjectProperty" key={key}>
                <span css={{ color: COLOR_GREEN }}>{key}</span>:{" "}
                <ValueReveal id={`${id}.${key}`} value={value[key]} />
              </div>
            ))}
          </div>
        )}
      </>
    );
  }

  if (typeof value === "string" && isUrlLike(value)) {
    return (
      <a
        data-testid="RecursiveReveal_Link"
        css={{
          color: "inherit",
          ":not(:hover)": {
            textDecoration: "none",
          },
        }}
        href={value}
        rel="noopener noreferrer"
        target="_blank"
      >
        {value}
      </a>
    );
  }

  // It's important to stringify the value, so strings are rendered with quotes..
  return <span data-testid="RecursiveReveal_Primitive">{JSON.stringify(value)}</span>;
}

export default function RecursiveReveal(props) {
  const [actives, setActives] = useState({});

  const toggle = useCallback(
    (id) => {
      const newActives = { ...actives };
      newActives[id] = !actives[id];

      setActives(newActives);
    },
    [actives],
  );

  const isActive = useCallback(
    (id) => {
      return !!actives[id];
    },
    [actives],
  );

  return (
    <Context.Provider value={{ isActive, toggle }}>
      <ValueReveal {...props} />
    </Context.Provider>
  );
}
