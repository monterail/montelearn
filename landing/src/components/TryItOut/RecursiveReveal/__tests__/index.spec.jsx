import { render, screen, fireEvent } from "@testing-library/react";

import RecursiveReveal from "../index";

describe("RecursiveReveal", () => {
  it("should render without errors", () => {
    expect(() => render(<RecursiveReveal value={null} />)).not.toThrow();
  });

  it("should render undefined", () => {
    render(<RecursiveReveal value={undefined} />);

    const element = screen.getByTestId("RecursiveReveal_Undefined");

    expect(element).toBeInTheDocument();
    expect(element).toMatchSnapshot();
  });

  it("should render null", () => {
    render(<RecursiveReveal value={null} />);

    const element = screen.getByTestId("RecursiveReveal_Null");

    expect(element).toBeInTheDocument();
    expect(element).toMatchSnapshot();
  });

  it("should render Date", () => {
    const date = new Date("2020-01-01T00:00:00.000Z");
    render(<RecursiveReveal value={date} />);

    const element = screen.getByTestId("RecursiveReveal_Date");

    expect(element).toBeInTheDocument();
    expect(element).toMatchSnapshot();
    expect(element.textContent).toBe(`Date (${date.toISOString()})`);
  });

  it("should render simple value", () => {
    render(<RecursiveReveal value={123} />);

    const element = screen.getByTestId("RecursiveReveal_Primitive");

    expect(element).toBeInTheDocument();
    expect(element).toMatchSnapshot();
  });

  it("should render empty array", () => {
    render(<RecursiveReveal value={[]} />);

    const element = screen.getByTestId("RecursiveReveal_Array");

    expect(element).toBeInTheDocument();
    expect(element).toMatchSnapshot();

    expect(element.textContent).toMatch(/^Array \(empty\)/);
  });

  it("should render empty object", () => {
    render(<RecursiveReveal value={{}} />);

    const element = screen.getByTestId("RecursiveReveal_Object");

    expect(element).toBeInTheDocument();
    expect(element).toMatchSnapshot();

    expect(element.textContent).toMatch(/^Object \(empty\)/);
  });

  describe("complex value", () => {
    const obj = {
      a: 1,
      b: [2, { c: 3 }],
    };

    it("should not render nested properties by default", () => {
      render(<RecursiveReveal value={obj} />);
      expect(screen.queryByTestId("RecursiveReveal_ObjectProperties")).not.toBeInTheDocument();
    });

    it("should reveal nested properties on click", () => {
      render(<RecursiveReveal value={obj} />);
      fireEvent.click(screen.getByTestId("RecursiveReveal_ObjectReveal"));
      expect(screen.queryByTestId("RecursiveReveal_ObjectProperties")).toBeInTheDocument();
    });

    it("should reveal and hide nested properties on clicks", () => {
      render(<RecursiveReveal value={obj} />);

      const element = screen.getByTestId("RecursiveReveal_ObjectReveal");

      fireEvent.click(element);
      expect(screen.queryByTestId("RecursiveReveal_ObjectProperties")).toBeInTheDocument();

      fireEvent.click(element);
      expect(screen.queryByTestId("RecursiveReveal_ObjectProperties")).not.toBeInTheDocument();
    });
  });
});
