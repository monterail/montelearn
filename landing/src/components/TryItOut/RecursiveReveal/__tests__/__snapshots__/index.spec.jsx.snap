// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecursiveReveal should render Date 1`] = `
<span
  data-testid="RecursiveReveal_Date"
>
  Date (
  2020-01-01T00:00:00.000Z
  )
</span>
`;

exports[`RecursiveReveal should render empty array 1`] = `
<span
  data-testid="RecursiveReveal_Array"
>
  <span
    class="css-avslql-ValueReveal"
  >
    Array
  </span>
   
  <i
    class="css-1k6t0vo-ValueReveal"
  >
    (
    empty
    )
  </i>
   
</span>
`;

exports[`RecursiveReveal should render empty object 1`] = `
<span
  data-testid="RecursiveReveal_Object"
>
  <span
    class="css-avslql-ValueReveal"
  >
    Object
  </span>
   
  <i
    class="css-1k6t0vo-ValueReveal"
  >
    (
    empty
    )
  </i>
   
</span>
`;

exports[`RecursiveReveal should render null 1`] = `
<span
  data-testid="RecursiveReveal_Null"
>
  NULL
</span>
`;

exports[`RecursiveReveal should render simple value 1`] = `
<span
  data-testid="RecursiveReveal_Primitive"
>
  123
</span>
`;

exports[`RecursiveReveal should render undefined 1`] = `
<span
  data-testid="RecursiveReveal_Undefined"
>
  undefined
</span>
`;
